stream.go

```diff
@@ -0,0 +1,451 @@
+	if err := bsb.UnmarshalRESP(br); err != nil {
+		return err
+	}
+
+	split := bytes.IndexByte(bsb.B, '-')
+	if split == -1 {
+		return errInvalidStreamID
+	}
+
+	time, err := bytesutil.ParseUint(bsb.B[:split])
+	if err != nil {
+		return errInvalidStreamID
+	}
+
+	seq, err := bytesutil.ParseUint(bsb.B[split+1:])
+	if err != nil {
+		return errInvalidStreamID
+	}
+
+	s.Time, s.Seq = time, seq
+	return nil
+}
+
+var _ fmt.Stringer = (*StreamEntryID)(nil)
+
+// String returns the ID in the format <time>-<seq> (the same format used by Redis).
+//
+// String implements the fmt.Stringer interface.
+func (s StreamEntryID) String() string {
+	var buf [maxUint64Len*2 + 1]byte
+	b := strconv.AppendUint(buf[:0], s.Time, 10)
+	b = append(b, '-')
+	b = strconv.AppendUint(b, s.Seq, 10)
+	return string(b)
+}
+
+// StreamEntry is an entry in a Redis stream as returned by XRANGE, XREAD and XREADGROUP.
+type StreamEntry struct {
+	// ID is the ID of the entry in a stream.
+	ID StreamEntryID
+
+	// Fields contains the fields and values for the stream entry.
+	Fields map[string]string
+}
+
+var _ resp.Unmarshaler = (*StreamEntry)(nil)
+
+var errInvalidStreamEntry = errors.New("invalid stream entry")
+
+// UnmarshalRESP implements the resp.Unmarshaler interface.
+func (s *StreamEntry) UnmarshalRESP(br *bufio.Reader) error {
+	var ah resp2.ArrayHeader
+	if err := ah.UnmarshalRESP(br); err != nil {
+		return err
+	}
+	if ah.N != 2 {
+		return errInvalidStreamEntry
+	}
+
+	if err := s.ID.UnmarshalRESP(br); err != nil {
+		return err
+	}
+
+	return s.unmarshalFields(br)
+}
+
+func (s *StreamEntry) unmarshalFields(br *bufio.Reader) error {
+	// resp2.Any{I: &s.Fields}.UnmarshalRESP(br)
+	var ah resp2.ArrayHeader
+	if err := ah.UnmarshalRESP(br); err != nil {
+		return err
+	}
+	if ah.N%2 != 0 {
+		return errInvalidStreamEntry
+	}
+
+	if s.Fields == nil {
+		s.Fields = make(map[string]string, ah.N/2)
+	} else {
+		for k := range s.Fields {
+			delete(s.Fields, k)
+		}
+	}
+
+	var bs resp2.BulkString
+	for i := 0; i < ah.N; i += 2 {
+		if err := bs.UnmarshalRESP(br); err != nil {
+			return err
+		}
+		key := bs.S
+		if err := bs.UnmarshalRESP(br); err != nil {
+			return err
+		}
+		s.Fields[key] = bs.S
+	}
+	return nil
+}
+
+// StreamReaderOpts contains various options given for NewStreamReader that influence the behaviour.
+//
+// The only required field is Streams.
+type StreamReaderOpts struct {
+	// Streams must contain one or more stream names that will be read.
+	//
+	// The value for each stream can either be nil or an existing ID.
+	// If a value is non-nil, only newer stream entries will be returned.
+	Streams map[string]*StreamEntryID
+
+	// Group is an optional consumer group name.
+	//
+	// If Group is not empty reads will use XREADGROUP with the Group as consumer group instead of XREAD.
+	Group string
+
+	// Consumer is an optional consumer name for use with Group.
+	//
+	// TODO(nussjustin): Automatically detect if empty and Group is set?
```

Comment:
Currently this defaults to an empty string, but we could try to automatically set something. Not sure if it's worth it.